_source/_posts/2023-09-15-java-microservices-spring.md

```diff
@@ -0,0 +1,679 @@
+```
+
+Then, I created an API in Auth0 called `fast-expiring-api` and set the TTL to 30 seconds.
+
+```shell
+auth0 apis create --name fast-expiring --identifier http://fast-expiring-api \
+  --token-lifetime 30
+```
+
+// todo: why does it prompt for scopes and offline access
+
+Restart the API gateway and go to `http://localhost:8080/print-token` to see your access token. You can copy the expired time to [timestamp-converter.com](https://www.timestamp-converter.com/) to see when it expires in your local timezone. Wait 30 seconds and refresh the page. You'll see a request to get a new token and an updated expires timestamp in your terminal.
+
+## The Okta Spring Boot starter and Keycloak
+
+If you find yourself in a situation where you don't have an internet connection, it can be handy to run Keycloak locally in a Docker container. Since the Okta Spring Boot starter is a thin wrapper around Spring Security, it works with Keycloak too. It does validate the issuer to make sure it's an Okta URL, so you have to use Spring Security's properties instead of the `okta.oauth2.*` properties.
+
+An easy way to get a pre-configured Keycloak instance is to use a [JHipster](https://www.jhipster.tech)'s `jhipster-sample-app-oauth2` application. It gets updated with every JHipster release. You can clone it with the following command:
+
+```shell
+git clone https://github.com/jhipster/jhipster-sample-app-oauth2.git --depth=1
+cd jhipster-sample-app-oauth2
+```
+
+Then, start Keycloak with `docker-compose -f src/main/docker/keycloak.yml up -d`. 
+
+You can configure the `api-gateway` to use Keycloak by removing the `okta.oauth2.*` properties and adding the following:
+
+```properties
+spring.security.oauth2.client.provider.okta.issuer-uri=http://localhost:9080/realms/jhipster
+spring.security.oauth2.client.registration.okta.client-id=web_app
+spring.security.oauth2.client.registration.okta.client-secret=web_app
+spring.security.oauth2.client.registration.okta.scope=openid,profile,email,offline_access
+```
+
+The `car-service` requires similar changes in its `application.properties` file:
+
+```properties
+spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:9080/realms/jhipster
+spring.security.oauth2.resourceserver.jwt.audiences=account
+```
+
+Restart both apps, open `http://localhost:8080`, and you'll be able to log in with Keycloak. 
+
+[Keycloak login screenshot]
+
+Use `admin`/`admin` for credentials and you'll be able to access `http://localhost:8080/cool-cars` as you did before.
+
+## Have fun with Spring Boot and Spring Cloud!
+
+I hope you liked this tour of how to build Java microservice architectures with Spring Boot and Spring Cloud. You learned how to build everything with minimal code, then configure it to be secure with Spring Security, OAuth 2.0, and Auth0 by Okta.
+
+You can find all the code shown in this tutorial [on GitHub](https://github.com/oktadev/auth0-java-microservices-examples/tree/main/spring-boot-gateway-webflux).
```

Comment:
TODO: Add links to related posts, as well as these labs:

- Auth0 CLI lab (or maybe do it inline)
- Spring Boot AuthN and AuthZ labs
- Social, MFA, and passkeys labs for workshop